### PR TITLE
core: fallback to unsafe logic of handling PID files

### LINKED ISSUE #8085: stricter PIDfile handling breaks several daemons
### Submission type

 - Bug report

### systemd version the issue has been seen with

v237

### Used distribution

Debian sid
Also reported downstream  as https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=889144

Quoting from the downstream bug report:

> The upstream commit db256aab13d8a89d583ecd2bacf0aca87c66effc "core: be 
stricter when handling PID files and MAINPID sd_notify() messages"
breaks several daemons in Debian.

> Known issues exist for

 > - munin-node https://bugs.debian.org/889073
 > - ulogd2
 > - dnsmasq https://lists.debian.org/debian-user/2018/01/msg01331.html

> and possibly others.

> Symptom is a timeout during service start, constant service restarts (if
> configured) and log messages like:

> Feb  2 14:22:49 HOST systemd[1]: ulogd2.service: Permission denied while opening PID file or unsafe > symlink chain: /run/ulog/ulogd.pid
> Feb  2 14:23:54 HOST systemd[1]: munin-node.service: Permission denied while opening PID file or unsafe symlink chain: /run/munin/munin-node.pid

In all those cases, a directory /run/foo/ is created, which is not owned by root (but a system user foo), the pid file that is written by the daemon is owned by root:root though. The pid file is not a symlink though.

The daemon in those cases drops privileges on its own and places additional files in that /run/foo directory, afaiu. Which is why the directory has to be owned by foo, so the unprivileged process can actually place files there.

As the pid file is not a symlink, I think systemd is incorrectly rejecting it.

### PR DESCRIPTION
**Pull request by yuwata**
The commit db256aab13d8a89d583ecd2bacf0aca87c66effc makes the PID-files handling logic stricter. This is desired, but breaks several daemons which unsafely manage their PID files. So, let's relax the restriction.
This should be reverted in the future release.

Fixes #8085.

----
I have not checked the tests provided by db256aab13d8a89d583ecd2bacf0aca87c66effc, and it may be necessary to update them. Let's wait the CIs.

### CODE REVIEW

File: src/core/service.c [hunk 1]
```diff
@@ -929,8 +929,11 @@ static int service_load_pid_file(Service *s, bool may_warn) {
         prio = may_warn ? LOG_INFO : LOG_DEBUG;
 
         fd = chase_symlinks(s->pid_file, NULL, CHASE_OPEN|CHASE_SAFE, NULL);
-        if (fd == -EPERM)
-                return log_unit_full(UNIT(s), prio, fd, "Permission denied while opening PID file or unsafe symlink chain: %s", s->pid_file);
+        if (fd == -EPERM) {
+                log_unit_full(UNIT(s), prio, fd, "Permission denied while opening PID file or unsafe symlink chain. Fallback to unsafe logic: %s", s->pid_file);
+                /* The following fallback logic is not safe. It may be dropped in the future release. */
+                fd = chase_symlinks(s->pid_file, NULL, CHASE_OPEN, NULL);
```
Comment (poettering) [inline src/core/service.c:935]: hmm, this fallback is acceptable, but if we do this, then we should enable the more restrictive in-cgroup checking for the PID still. i.e. we should set a bool here, that we check further down when service_is_suitable_main_pid() returns 0: in that case we should *not* accept the PID file if it points to a PID outside of the cgroup.

Explanation: service_is_suitable_main_pid() returns < 0 if the specified PID is not suitable as main PID for the service. It returns > 0 if it is fully safe (i.e. within the service's cgroup). But it reports == 0 if its "questionnable", i.e. not within the cgroup itself. In that case we currently do a simple fstat() on the PID file and if its owned by root accept it either way. But precisely *that* should be bypassed if the symlink was unsafe: we shouldn't allow questionnable PIDs from PID files in unsafe symlink chains even if the final file is owned by root...

File: src/core/service.c [hunk 2]
```diff
@@ -929,8 +929,11 @@ static int service_load_pid_file(Service *s, bool may_warn) {
         prio = may_warn ? LOG_INFO : LOG_DEBUG;
 
         fd = chase_symlinks(s->pid_file, NULL, CHASE_OPEN|CHASE_SAFE, NULL);
-        if (fd == -EPERM)
-                return log_unit_full(UNIT(s), prio, fd, "Permission denied while opening PID file or unsafe symlink chain: %s", s->pid_file);
+        if (fd == -EPERM) {
+                log_unit_full(UNIT(s), prio, fd, "Permission denied while opening PID file or unsafe symlink chain. Fallback to unsafe logic: %s", s->pid_file);
```
Comment (poettering) [inline src/core/service.c:933]: we should probably have this is recognizable error that carries an explanation in the message catalog. i.e. use log_struct(…,  "MESSAGE_ID=" …

**Comment by yuwata** [comment]
@poettering Thank you for the comment. Tentatively, I close this to not disturb anyone try to fix the bug(?) and implement the logic as you commented. For me, it is a little bit tough work right now.

**Comment by poettering** [comment]
Prepped the suggested fix in #8133 now. @mbiebl could you check if that makes things work for you?

**Pull request closed without merging**